src/SFA.DAS.AssessorService.Domain/Paging/PaginatedList.cs

```diff
@@ -3,27 +3,66 @@
 {
-    public class PaginatedList<T>
+    public abstract class PaginatedList
     {
-        public int PageIndex { get; set; }
-        public int PageSize { get; set; }
-        public int TotalRecordCount { get; set; }
+        public abstract int PageIndex { get; }
+        public abstract int PageSize { get; }
+        public abstract int PageSetSize { get; set; }
+
+        public abstract int TotalRecordCount { get; }
+        public abstract int TotalPages { get; }
+
+        public abstract bool HasPreviousPage { get; }
+        public abstract bool HasNextPage { get; }
+
+        public abstract int FirstVisiblePage { get; }
+        public abstract int LastVisiblePage { get; }
+
+        public abstract int FirstVisibleItem { get; }
+        public abstract int LastVisibleItem { get; }
+    }
+
+    public class PaginatedList<T> : PaginatedList
+    {
+        public override int PageIndex { get; }
+        public override int PageSize { get; }
+        public override int PageSetSize { get; set; }
 
         public List<T> Items { get; } = new List<T>();
 
-        public int TotalPages { get; set; }
+        public override int TotalRecordCount { get; }
+        public override int TotalPages => (TotalRecordCount / PageSize) + ((TotalRecordCount % PageSize) > 0 ? 1 : 0);
 
-        public PaginatedList(List<T> items, int count, int pageIndex, int pageSize)
+        private PaginatedList(List<T> items, int totalRecordCount, int pageIndex, int pageSize)
         {
             PageIndex = pageIndex;
             PageSize = pageSize;
-            TotalRecordCount = count;
-            TotalPages = (int)Math.Ceiling(count / (double)pageSize);
+            TotalRecordCount = totalRecordCount;
+
+            if (items != null)
+            {
+                Items.AddRange(items);
+            }
+        }
+
+        public PaginatedList(List<T> items, int totalRecordCount, int pageIndex, int pageSize, int? pageSetSize = null)
+            : this(items, totalRecordCount, pageIndex, pageSize)
+        {
+            PageSetSize = pageSetSize ?? 1;
+        }
 
-            if(items != null) Items.AddRange(items);
+        public override bool HasPreviousPage => (PageIndex > 1);
+        public override bool HasNextPage => (PageIndex < TotalPages);
+
+        public override int FirstVisiblePage => Math.Max(Math.Min(PageIndex - (PageSetSize / 3), (TotalPages + 1) - PageSetSize), 1);
+        public override int LastVisiblePage => Math.Min(Math.Max(PageIndex + (PageSetSize / 2), PageSetSize), TotalPages);
+
+        public override int FirstVisibleItem => ((PageIndex-1) * PageSize) + 1;
+        public override int LastVisibleItem => Math.Min(PageIndex * PageSize, TotalRecordCount);
+
+        public PaginatedList<T1> Convert<T1>() where T1 : class
+        {
+            return new PaginatedList<T1>(Items.ConvertAll(p => p as T1), TotalRecordCount, PageIndex, PageSize, PageSetSize);
         }
```

Comment:
This code is never used

Reply:
This is a copy of the code from Assessor which does use it - I was trying to keep it the same - as really this whole project belongs in the Assessor Repo - there are copies of the code which are 99.5% the same between the repos